src/vs/workbench/services/editor/common/customEditorLabelService.ts

```diff
@@ -115,7 +119,18 @@ export class CustomEditorLabelService extends Disposable implements ICustomEdito
 		if (!this.enabled) {
 			return undefined;
 		}
-		return this.applyPatterns(resource);
+
+		const cached = this.cache.get(resource);
+		if (cached !== undefined) {
+			return cached;
+		}
+
+		const result = this.applyPatterns(resource);
+		if (result !== undefined) {
```

Comment:
I would add it to the cache even when the result of the computation is undefined, so the next time it checks the file path it knows there is no matching pattern and does not have to do any computations